types/wlr_pointer_constraints_v1.c

```diff
@@ -0,0 +1,387 @@
+	}
+	constraint->accepted = true;
+	pointer_constraint_check_active(constraint);
+}
+
+void wlr_pointer_constraint_v1_decline(struct wlr_pointer_constraint_v1 *constraint) {
+	wlr_log(L_DEBUG, "constraint %p declined", constraint);
+	constraint->accepted = false;
+	send_unconstrained(constraint);
+}
+
+static void resource_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static void pointer_constraint_destroy(struct wlr_pointer_constraint_v1 *constraint) {
+	if (constraint == NULL) {
+		return;
+	}
+
+	if (constraint->seat_client->seat->current_pointer_constraint == constraint) {
+		constraint->seat_client->seat->current_pointer_constraint = NULL;
+	}
+
+	send_unconstrained(constraint);
+	wl_resource_set_user_data(constraint->resource, NULL);
+	wl_list_remove(&constraint->link);
+	wl_list_remove(&constraint->surface_commit.link);
+	wl_list_remove(&constraint->surface_destroy.link);
+	free(constraint);
+}
+
+static void pointer_constraint_destroy_resource(struct wl_resource *resource) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	pointer_constraint_destroy(constraint);
+}
+
+static void pointer_constraint_set_region(struct wl_client *client,
+		struct wl_resource *resource, struct wl_resource *region) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.region = region;
+}
+
+// only valid for locked_pointer
+static void pointer_constraint_set_cursor_position_hint(struct wl_client *client,
+		struct wl_resource *resource, wl_fixed_t x, wl_fixed_t y) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		pointer_constraint_from_resource(resource);
+
+	constraint->pending.cursor_hint.x = wl_fixed_to_double(x);
+	constraint->pending.cursor_hint.y = wl_fixed_to_double(y);
+	constraint->pending.cursor_hint.valid = true;
+}
+
+static void handle_surface_commit(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_commit);
+
+	if (constraint->current.region != constraint->pending.region) {
+		send_unconstrained(constraint);
+	}
+	constraint->current = constraint->pending;
+}
+
+static void handle_surface_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_pointer_constraint_v1 *constraint =
+		wl_container_of(listener, constraint, surface_destroy);
+
+	send_unconstrained(constraint);
+	wl_resource_destroy(constraint->resource); // is this needed?
+}
+
+static const struct zwp_confined_pointer_v1_interface confined_pointer_impl = {
+	.destroy    = resource_destroy,
+	.set_region = pointer_constraint_set_region,
+};
+
+static const struct zwp_locked_pointer_v1_interface locked_pointer_impl = {
+	.destroy                  = resource_destroy,
+	.set_cursor_position_hint = pointer_constraint_set_cursor_position_hint,
+	.set_region               = pointer_constraint_set_region,
```

Comment:
We don't do alignment like this, just do one space on either side of `=` and call it a day